Enhance: Bring back user to 1st page when filters are changed

imanish003 opened the pull request:
**URL page number removal on filter change**: When filters are changed, the page number in the URL is removed. This is to ensure that when a filter is applied, the user is taken back to the first page of results. The code accommodates three different formats of page numbers in the URL and removes them accordingly.

Part of #9876 

### Testing

1. Create a new post.
2. Add all filter blocks and a 'Products Beta' block to it.
3. Save the post and navigate to the frontend.
4. Navigate to the 2nd page using pagination. 
5. Modify any filter and confirm that the page number resets to the 1st page upon filter alteration. 
6. Repeat the above steps and validate that this also holds true for:
    - 'Product Collection' block (The product collection block is currently in the experimental phase and is not shipped with the plugin/core)
    - 'Classic WooCommerce Product Grid Block'
    - Note: This behavior doesn't affect the 'All Products' block.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [X] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Enhancement: Bring back user to 1st page when filters are changed

imanish003 committed:
Add support for filter blocks

This commit updates the `ProductCollection` block implementation to add support for filter blocks, including the Price filter, Attributes filter, Rating filter, and In stock filter.

The changes include:

1. Adding a new property `$attributes_filter_query_args` to store the query arguments related to the filter by attributes block.

2. Adding a new method `add_support_for_filter_blocks()` to enable support for filter blocks. This method sets the necessary asset data to enable filtering and refreshes the page when a filter is applied.

3. Adding new methods to handle specific filter queries, including `get_filter_by_price_query()`, `get_filter_by_attributes_query()`, `get_filter_by_stock_status_query()`, and `get_filter_by_rating_query()`. These methods generate the respective queries based on the applied filters.

4. Refactoring the `get_final_query_args()` method to include the newly added filter queries using the `get_queries_by_applied_filters()` method.

These changes enhance the functionality of the `ProductCollection` block by allowing users to filter products based on price, attributes, rating, and stock status.

imanish003 committed:
Go to first page when filters are updated

imanish003 committed:
Enhance ProductCollection block to support filter blocks

This commit enhances the ProductCollection block to support various filter blocks such as Price filter block, Attributes filter block, Rating filter block, and In stock filter block.

The `build_query` method has been refactored into two separate methods: `build_frontend_query` and `get_final_frontend_query` to make the code more modular and readable. The `add_support_for_filter_blocks` method has been modified to support the generation of product IDs for filter blocks.

The method `update_rest_query` has been renamed to `update_rest_query_in_editor` for better clarity and understanding of its function. Similarly, `get_final_query_args` has been refactored to include the `$is_exclude_applied_filters` parameter which helps in generating product IDs for the filter blocks.

Moreover, the filter hook `pre_render_block` has been added to support the filtering of blocks before they are rendered.

This update will enhance the user experience by providing more filtering options in the ProductCollection block.

imanish003 committed:
Merge branch 'trunk' into fix/9876-product-collection-should-work-well-with-filter-blocks

imanish003 committed:
Merge branch 'trunk' into fix/9876-product-collection-should-work-well-with-filter-blocks

imanish003 committed:
Merge branch 'trunk' into fix/9876-product-collection-should-work-well-with-filter-blocks

imanish003 committed:
Remove changes related to redirect to 1st page

imanish003 committed:
Go to first page when filters are updated

This commit refactors the changeUrl function in the `filters.ts` file. The purpose of this refactor is to remove the page number from the URL whenever filters are changed. This ensures that the user is always taken to the first page of results.

The commit includes the following changes:
- Added comments explaining the different page number formats in the URL.
- Updated the `newUrl` parameter by removing the page number using regular expressions.
- The page is then reloaded with the updated URL using `window.location.href`.

These changes improve the URL handling in the `changeUrl` function and enhance the user experience when interacting with filters.

imanish003 committed:
Merge branch 'trunk' into fix/9876-product-collection-should-work-well-with-filter-blocks-2

kmanijak reviewed (approved):
LGTM!

imanish003 committed:
Merge branch 'trunk' into fix/9876-product-collection-should-work-well-with-filter-blocks-2

Aljullu reviewed (commented):
I noticed that when removing all filters, the URL becomes `/?` first and then `/`, causing an unnecessary reload:

[Enregistrament de pantalla des de 2023-06-22 14-56-46.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/bc3a44fa-7474-48a7-9ae3-df22a830a2b0)

Can you reproduce as well?

imanish003 commented:
@Aljullu, I can reproduce the issue. Thank you for identifying this edge case. I will work on fixing it and push the necessary changes 🤝

imanish003 committed:
Optimize URL updates to prevent unnecessary page reloads

This commit optimizes the `changeUrl` utility function to prevent unnecessary page reloads.

Changes:
1. It checks if there are no query parameters in the URL and the URL ends with a '?'.
2. If these conditions are met, it removes the trailing '?' from the URL.

The reason for this change is that a '?' in the URL typically indicates that query parameters will follow. If there are no query parameters, the '?' is unnecessary and can cause the page to reload, which can negatively affect performance. Therefore, by removing the '?', we prevent this unnecessary page reload.

imanish003 commented:
Hi @Aljullu, I have pushed a fix in [872c583](https://github.com/woocommerce/woocommerce-blocks/pull/9952/commits/872c583b3d868ee6795566820db87faa00e8d404). Can you please take a look again? 🙂

Aljullu commented on assets/js/utils/filters.ts:
```diff
@@ -50,7 +50,19 @@ export function changeUrl( newUrl: string ) {
+		 * performance. Therefore, by removing the '?', we prevent this unnecessary reload.
+		 */
+		const urlObject = new URL( newUrl );
+		const numberOfQueryParams = [ ...urlObject.searchParams.keys() ].length;
+		if ( numberOfQueryParams === 0 && newUrl.endsWith( '?' ) ) {
```
I think it's safe to skip the query params check (`numberOfQueryParams === 0`), no? I mean, if the URL ends with `?`, it's always safe to remove the last question mark, no matter if there are query params or not, no? :thinking:

imanish003 replied:
I think so. I was just trying to be extra careful to ensure I do not break any existing functionality. 

But good point, the trailing "?" can indeed be removed safely without checking for query parameters. I will simplify the code accordingly. Thanks for the suggestion! 🙌🏻

imanish003 replied:
@Aljullu I have made the changes in [9a82cd9](https://github.com/woocommerce/woocommerce-blocks/pull/9952/commits/9a82cd95a27a1031d9e7b78475b334f812e1cf8b). Can you please take a look again just to be sure that everything still works as expected? 🙂

Aljullu reviewed (approved):
This is working great, @imanish003! I just left one small comment about a possible small optimization, but besides that, LGTM.

imanish003 committed:
Merge branch 'trunk' into fix/9876-product-collection-should-work-well-with-filter-blocks-2

imanish003 committed:
Simplify trailing '?' removal in URL handling

This commit simplifies the removal of trailing '?' characters from URLs.
The previous implementation checked if there were any query parameters
before removing a trailing '?'.

However, upon review, it was determined that this check was unnecessary,
as a trailing '?' could be removed safely even if query parameters were
present, without any negative impact.

This change simplifies the code and prevents unnecessary computation
checking for the existence of query parameters.

Aljullu reviewed (commented):
Thanks for the update, @imanish003. I just found one last minor issue: in some occassions I end up with two `&&` in the URL.

Steps to reproduce:

1. Add the Filter by Price and Filter by Attribute blocks alongside the Product Collection block.
2. In the frontend, add a price filter so pagination is still available. The URL will look something like `/?min_price=15`.
3. Using pagination, switch to page two. The URL will look something like `/?query-0-page=2&min_price=15`.
4. Now, add an attribute filter. The URL will look something like `/?min_price=15&&filter_color=blue&query_type_color=or`. Notice the double `&&` before `filter_color`.

Can you reproduce as well?

imanish003 commented:
Hi @Aljullu, This issue isn't reproducing for me 🤔  Here is a quick demo of what I tried:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/698003b8-46da-4cbe-8d23-cfb542a14fd6

If you are still able to reproduce the issue then maybe I can add a logic like this to clean up query params:
```
let url = new URL(newUrl);
let params = new URLSearchParams(url.search);

// Assign the cleaned up search parameters back to the URL
url.search = params.toString();

// Now update newUrl with the cleaned up URL
newUrl = url.toString();
```
What do you think? 🙂

Aljullu reviewed (approved):
Ups, I don't seem to be able to reproduce anymore, sorry for the confusion. :confused: Not sure what was wrong with my setup, I will open an issue if I can reproduce again, but for now, let's merge this PR. :rocket:
